backend/src/apis/csv_parser.rs

```diff
@@ -1,49 +1,118 @@
-use std::error::Error;
-use std::fs::File;
-use std::path::Path;
+use std::{error::Error, fs::File, path::Path};
 
+use ethers::{
+    abi::AbiEncode,
+    types::{Bytes, U256},
+};
 use serde::Deserialize;
 
+use crate::contracts::generated::summa_contract::{AddressOwnershipProof, Asset};
+
 #[derive(Debug, Deserialize)]
-struct Record {
+struct EntriesRecord {
+    chain: String,
     address: String,
     signature: String,
+    message: String,
 }
 
 pub fn parse_signature_csv<P: AsRef<Path>>(
     path: P,
-) -> Result<(Vec<String>, Vec<String>), Box<dyn Error>> {
+) -> Result<Vec<AddressOwnershipProof>, Box<dyn Error>> {
+    let file = File::open(path)?;
+    let mut rdr = csv::ReaderBuilder::new().delimiter(b';').from_reader(file);
+
+    let mut address_ownership_proofs = Vec::<AddressOwnershipProof>::new();
+
+    for result in rdr.deserialize() {
+        let record: EntriesRecord = result?;
+
+        address_ownership_proofs.push(AddressOwnershipProof {
+            cex_address: record.address.to_string(),
+            chain: record.chain.to_string(),
+            signature: record.signature.parse()?,
+            message: Bytes::from(record.message.encode()),
+        });
+    }
+
+    Ok(address_ownership_proofs)
+}
+
+#[derive(Debug, Deserialize)]
+struct AssetRecord {
+    chain: String,
+    asset_name: String,
+    amount: String,
+}
+
+pub fn parse_asset_csv<P: AsRef<Path>, const N_ASSETS: usize>(
+    path: P,
+) -> Result<[Asset; N_ASSETS], Box<dyn Error>> {
     let file = File::open(path)?;
     let mut rdr = csv::ReaderBuilder::new().delimiter(b';').from_reader(file);
 
-    let mut signatures = Vec::<String>::new();
-    let mut addresses = Vec::<String>::new();
+    let mut assets_vec = Vec::with_capacity(N_ASSETS);
 
     for result in rdr.deserialize() {
-        let record: Record = result?;
+        let record: AssetRecord = result?;
 
-        signatures.push(record.signature);
-        addresses.push(record.address);
+        assets_vec.push(Asset {
+            asset_name: record.asset_name,
+            chain: record.chain,
+            amount: U256::from_dec_str(&record.amount)?,
+        });
     }
 
-    Ok((addresses, signatures))
+    let assets_array: [Asset; N_ASSETS] = assets_vec.try_into().map_err(|v: Vec<Asset>| {
+        format!(
+            "Number of assets in CSV does not match the number of assets in the contract! {:?}",
```

Comment:
The error message is misleading, it should say something like "the number of assets doesn't match the expected"

Reply:
Updated at [31c3c59](https://github.com/summa-dev/summa-solvency/pull/152/commits/31c3c59bf0d5efa6027b238ee7bb54bceb1b8800)